### PR TITLE
LUCENE-9572 - TypeAsSynonymFilter gains selective flag transfer and an ignore list.

### PR DESCRIPTION
**Pull request by gus-asf**

### CODE REVIEW

**Comment by gus-asf** [comment]
@sarowe do you have any comments?

**Comment by sarowe** [comment]
> @sarowe do you have any comments?

I'll take a look later today.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilter.java [hunk 1]
```diff
@@ -64,9 +82,15 @@ public boolean incrementToken() throws IOException {
       }
       termAtt.append(typeAtt.type());
       posIncrAtt.setPositionIncrement(0);
+      // control what flags transfer to synonym
+      int flags = getAttribute(FlagsAttribute.class).getFlags();
```
Comment (uschindler) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilter.java:86]: This should not call getAttribute and instead use the instance `flagsAtt` above (instance field). In the next line it's correct.

File: lucene/CHANGES.txt [hunk 2]
```diff
@@ -15,7 +15,7 @@ API Changes
 * LUCENE-8474: RAMDirectory and associated deprecated classes have been
   removed. (Dawid Weiss)
 
-* LUCENE-3041: The deprecated Weight#extractTerms() method has been 
+* LUCENE-3041: The deprecated Weight#extractTerms() method has been
```
Comment (uschindler) [inline lucene/CHANGES.txt:18]: same issue like in previous PR!

**Review by uschindler — CHANGES_REQUESTED** [review]
This looks fine to me except the "getAttribute". It should only use flagsAtt.
I don't fully understand why this flags handling is needed in the tokenfilter, because i have not yet reviewed the new Solr feature.
At least I cannot find any problematic parts. To me it's mostly important if the tokenfilter works the same like before if no exclusion set and no flags are passed, which seems guaranteed by `~0` (cool notation for all bits set)!

File: lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java [hunk 3]
```diff
@@ -190,7 +196,8 @@ public static void assertTokenStreamContents(TokenStream ts, String[] output, in
       if (posLengthAtt != null) posLengthAtt.setPositionLength(45987653);
       if (keywordAtt != null) keywordAtt.setKeyword((i&1) == 0);
       if (payloadAtt != null) payloadAtt.setPayload(new BytesRef(new byte[] { 0x00, -0x21, 0x12, -0x43, 0x24 }));
-      
+      if (flagsAtt != null) flagsAtt.setFlags(Integer.MAX_VALUE); // all 1's
```
Comment (uschindler) [inline lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java:199]: that's all bits except the first 1 (`01111111111111111...`). To invert all bits use `-1` or better `~0`

File: lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java [hunk 4]
```diff
@@ -294,6 +304,7 @@ public static void assertTokenStreamContents(TokenStream ts, String[] output, in
     if (posLengthAtt != null) posLengthAtt.setPositionLength(45987653);
     if (keywordAtt != null) keywordAtt.setKeyword(true);
     if (payloadAtt != null) payloadAtt.setPayload(new BytesRef(new byte[] { 0x00, -0x21, 0x12, -0x43, 0x24 }));
+    if (flagsAtt != null) flagsAtt.setFlags(Integer.MAX_VALUE); // all 1's
```
Comment (uschindler) [inline lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java:307]: same here

File: lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java [hunk 5]
```diff
@@ -125,7 +125,7 @@ public void reflectWith(AttributeReflector reflector) {
   //     arriving to pos Y have the same endOffset)
   public static void assertTokenStreamContents(TokenStream ts, String[] output, int startOffsets[], int endOffsets[], String types[], int posIncrements[],
```
Comment (uschindler) [inline lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java:126]: maybe we should not change this method signature and just add another one, so we do not need to change unrelated files like MinHashFilter?

**Review by uschindler — CHANGES_REQUESTED** [review]

File: lucene/CHANGES.txt (hunk 2 above)
Comment (gus-asf) [reply lucene/CHANGES.txt:18]: Well actually this PR predates other :), but yes same issue.

File: lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java (hunk 3 above)
Comment (gus-asf) [reply lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java:199]: oops yeah realized that partway through (when I switched to ~0 for the other file), didn't come back and fix it here.

File: lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java (hunk 5 above)
Comment (gus-asf) [reply lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java:126]: Done, I think what was originally submitted was a bit easier than sorting out all the parameters at the time, but yes nicer to minimize.

**Comment by gus-asf** [comment]
> I don't fully understand why this flags handling is needed in the tokenfilter, because i have not yet reviewed the new Solr feature.

LUCENE-9575 (patch waiting for the test changes here) will show you where the flags are gained. The basic idea is that the flags are used to identify the source tokens that gave rise to synonym tokens and (sometimes) remove the source tokens *after* they participate in further (arbitrary) processing... the canonical example is to allow '401k' and '401(k)' to match the same documents about retirement plans and also NOT match documents regarding "401 Unauthorized" or "Vitamin K"

**Review by uschindler — APPROVED** [review]
Looks fine to me. I am waiting to see an example how this is used, maybe with the Solr example you bring later.

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/miscellaneous/TestTypeAsSynonymFilter.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  public void testWithPrefixAndIgnore() throws Exception {
+    Token[] tokens = new Token[] {
+        new Token("foo", 1, 3),
+        new Token("foo", 5, 7),
+        new Token("foo", 9, 11),
+    } ;
+    tokens[0].setType("bar");
+    tokens[2].setType("ignoreme");
+    TokenStream ts = new CannedTokenStream(tokens);
+    ts = new TypeAsSynonymFilter(ts,"pfx_", Stream.of("word","ignoreme").collect(Collectors.toSet()), 0);
```
Comment (uschindler) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/miscellaneous/TestTypeAsSynonymFilter.java:68]: Why not use `Set.of()` (since Java 11)?

Comment (uschindler) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/miscellaneous/TestTypeAsSynonymFilter.java:68]: `Stream.of("word","ignoreme").collect(Collectors.toSet())` => `Set.of("word","ignoreme")`

**Comment by uschindler** [comment]
I just recommend a small change in the test to use Set.of(), hwich is new in Java 11 (I think already in 9) for static sets.

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/miscellaneous/TestTypeAsSynonymFilter.java (hunk 6 above)
Comment (gus-asf) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/miscellaneous/TestTypeAsSynonymFilter.java:68]: Ah yeah sure. Originally developed vs 8.4 but now I can use it :)

Comment (uschindler) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/miscellaneous/TestTypeAsSynonymFilter.java:68]: In 8.x I'd use: `new HashSet(Arrays.asList("word","ignoreme"))`, but that's a matter of preference. I tend to not use streams if there's no transformations/mappings going on.

**Comment by gus-asf** [comment]
Puzzling why RulesTest in solr is suddenly failing rat. It sure looks like it has a license just like all others and I didn't touch it.

**Comment by gus-asf** [comment]
Well that was misleading. Rat is happy locally, not sure why github was blaming RulesTest's rat analysis. I did find a couple of unused imports, and realized that apparently `gradlew :lucene:check` doesn't run ecjLint as I had expected. Just for fun javadocs were running out of memory locally when I tried a full top level precommit which gives this lovely cryptic message: 
```
* What went wrong:
Execution failed for task ':lucene:demo:renderJavadoc'.
> java.io.IOException: Stream Closed
```
(had to edit line 543 of render-javadoc.gradle to get it to work)

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilter.java [hunk 7]
```diff
@@ -34,23 +36,39 @@
   private final CharTermAttribute termAtt = addAttribute(CharTermAttribute.class);
```
Comment (sarowe) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilter.java:36]: Consider adding the new features to the class comment

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilterFactory.java [hunk 8]
```diff
@@ -18,12 +18,15 @@
...
 import java.util.Map;
+import java.util.Set;
 
 import org.apache.lucene.analysis.TokenStream;
 import org.apache.lucene.analysis.TokenFilterFactory;
 
 /**
  * Factory for {@link TypeAsSynonymFilter}.
+ *
+ * <p>In Solr this might be used as such
```
Comment (sarowe) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilterFactory.java:29]: Consider adding configuration for the new features to the example Solr configuration.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilterFactory.java [hunk 9]
```diff
@@ -46,10 +49,17 @@
...
+  private final int synFlagMask;
 
   public TypeAsSynonymFilterFactory(Map<String,String> args) {
     super(args);
     prefix = get(args, "prefix");  // default value is null
+    String ignoreList = get(args, "ignore");
+    synFlagMask = getInt(args,"synFlagsMask", ~0);
+    if (ignoreList != null) {
+      ignore = Set.of(ignoreList.split(","));
+    }
```
Comment (sarowe) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilterFactory.java:62]: There is a method `getSet()` which you could use instead of `get()` for the ignoreList; it does the splitting for you, and also supports mixed space/comma separators.

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/miscellaneous/TestTypeAsSynonymFilter.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.lucene.analysis.BaseTokenStreamTestCase;
+import org.apache.lucene.analysis.CannedTokenStream;
+import org.apache.lucene.analysis.Token;
+import org.apache.lucene.analysis.TokenStream;
+
+/**
+ * Test that this filter moves the value in type to a synonym token with the same offsets. This is rarely
+ * useful by itself, but in combination with another filter that updates the type value with an appropriate
+ * synonym can be used to identify synonyms before tokens are modified by further analysis, and then
+ * add them at the end, ensuring that the synonym value has not ben subjected to the intervening analysis.
```
Comment (sarowe) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/miscellaneous/TestTypeAsSynonymFilter.java:31]: typo: ben

File: lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java [hunk 11]
```diff
@@ -390,19 +406,19 @@ public static void assertAnalyzesTo(Analyzer a, String input, String[] output, i
   }
 
   public static void assertAnalyzesTo(Analyzer a, String input, String[] output, int startOffsets[], int endOffsets[], String types[], int posIncrements[], int posLengths[], boolean graphOffsetsAreCorrect, byte[][] payloads) throws IOException {
-    assertTokenStreamContents(a.tokenStream("dummy", input), output, startOffsets, endOffsets, types, posIncrements, posLengths, input.length(), null, null, graphOffsetsAreCorrect, payloads);
+    assertTokenStreamContents(a.tokenStream("dummy", input), output, startOffsets, endOffsets, types, posIncrements, posLengths, input.length(), null, null, graphOffsetsAreCorrect, payloads, null);
```
Comment (sarowe) [inline lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java:409]: Since you put back the original method, I think you can also revert this change?

**Review by sarowe — CHANGES_REQUESTED** [review]

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilterFactory.java (hunk 9 above)
Comment (gus-asf) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilterFactory.java:62]: Ah cool.

File: lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java (hunk 11 above)
Comment (gus-asf) [reply lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamTestCase.java:409]: ok

**Comment by gus-asf** [comment]
In re-reading things, I think it just dawned on me that I'm not quite maintaining back compatibility after all. Previously no flags at all were transferred, yet my ~0 default now copies all flags. ( x & ~0 = x) but the prior functionality was no flags copied which would be ( x & 0 = 0) (plus we could test for 0 and just not set flags at all)

**Comment by gus-asf** [comment]
Oh darn, out thought myself looking at the diff where I added a setFlags.... forgot that the stored state is copying flags already never-mind.

**Comment by gus-asf** [comment]
@sarowe I think I've addressed your concerns LMK if you have further thoughts

**Comment by sarowe** [comment]
> @sarowe I think I've addressed your concerns LMK if you have further thoughts

Yes you have, with one exception: I left a comment that AFAICT you haven't addressed:

> .../TypeAsSynonymFilter.java
> Consider adding the new features to the class comment

**Comment by gus-asf** [comment]
Oh I added it to the Factory class which already has a solr xml example (https://github.com/apache/lucene-solr/pull/1965/files#diff-6efa63ada0f2520335d809359639f4c280a4ffbed3cac47927f2e47009f5f7f7R41)

**Comment by sarowe** [comment]
> Oh I added it to the Factory class which already has a solr xml example (https://github.com/apache/lucene-solr/pull/1965/files#diff-6efa63ada0f2520335d809359639f4c280a4ffbed3cac47927f2e47009f5f7f7R41)

Yeah I saw that, but the Filter is an entry point independent of the Factory, and its class comment contains a brief description of the filter: https://github.com/apache/lucene-solr/blob/master/lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/TypeAsSynonymFilter.java#L29-L32
.

I was thinking another phrase or two could be added mentioning the ignore list and the flag masking capabilities, but if you disagree, that's fine.

**Comment by gus-asf** [comment]
I don't disagree exactly, just seemed like that was where the doc was. Separately it might be good to have a policy that the doc is either on the factory or the filter and @ see to the other to avoid duplication. Happy to add a tidbit on the filter for now.

**Review by sarowe — APPROVED** [review]
